File: src/FSharpVSPowerTools/Commands/UnionPatternMatchCaseGeneratorSmartTaggerProvider.cs
using System;
using System.ComponentModel.Composition;
using Microsoft.VisualStudio.Text;
using Microsoft.VisualStudio.Text.Editor;
using Microsoft.VisualStudio.Text.Operations;
using Microsoft.VisualStudio.Text.Tagging;
using Microsoft.VisualStudio.Utilities;
using Microsoft.VisualStudio.Shell;
using System.Diagnostics;
using Microsoft.VisualStudio.Shell.Interop;
using FSharp.Editing.VisualStudio.CodeGeneration;
using FSharp.Editing.VisualStudio.ProjectSystem;
using FSharp.Editing.VisualStudio;

namespace FSharpVSPowerTools
{
    [Export(typeof(IViewTaggerProvider))]
    [ContentType("F#")]
    [TagType(typeof(UnionPatternMatchCaseGeneratorSmartTag))]
    [TextViewRole(PredefinedTextViewRoles.Editable)]
    [TextViewRole(PredefinedTextViewRoles.PrimaryDocument)]
    public class UnionPatternMatchCaseGeneratorSmartTaggerProvider : IViewTaggerProvider
    {
        private readonly IServiceProvider _serviceProvider;
        private readonly ITextDocumentFactoryService _textDocumentFactoryService;
        private readonly ProjectFactory _projectFactory;
        private readonly VSLanguageService _fsharpVsLanguageService;
        private readonly IVSOpenDocumentsTracker _openDocumentsTracker;
        private readonly ITextUndoHistoryRegistry _undoHistoryRegistry;
        
        [ImportingConstructor]
        public UnionPatternMatchCaseGeneratorSmartTaggerProvider(
            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
            ITextDocumentFactoryService textDocumentFactoryService,
            ITextUndoHistoryRegistry undoHistoryRegistry,
            ProjectFactory projectFactory,
            VSLanguageService fsharpVsLanguageService,
            IVSOpenDocumentsTracker openDocumentsTracker)
        {
            _serviceProvider = serviceProvider;
            _textDocumentFactoryService = textDocumentFactoryService;
            _undoHistoryRegistry = undoHistoryRegistry;
            _projectFactory = projectFactory;
            _fsharpVsLanguageService = fsharpVsLanguageService;
            _openDocumentsTracker = openDocumentsTracker;
        }

        public ITagger<T> CreateTagger<T>(ITextView textView, ITextBuffer buffer) where T : ITag
        {
            // Only provide the smart tagger on the top-level buffer
            if (textView.TextBuffer != buffer) return null;

            var generalOptions = Setting.getGeneralOptions(_serviceProvider);
            if (generalOptions == null || !generalOptions.UnionPatternMatchCaseGenerationEnabled) return null;
            var codeGenOptions = Setting.getCodeGenerationOptions(_serviceProvider);
            if (codeGenOptions == null) return null;

            var dte = _serviceProvider.GetService(typeof(SDTE)) as EnvDTE.DTE;
            var vsVersion = VisualStudioVersionModule.fromDTEVersion(dte.Version);
            if (vsVersion >= VisualStudioVersion.VS2015) return null;

            ITextDocument doc;
            if (_textDocumentFactoryService.TryGetTextDocument(buffer, out doc))
            {
                var generator = 
                    new UnionPatternMatchCaseGenerator(
                        doc, 
                        textView,
                        _undoHistoryRegistry.RegisterHistory(buffer),
                        _fsharpVsLanguageService,
                        _projectFactory, 
                        Setting.getDefaultMemberBody(codeGenOptions),
                        _openDocumentsTracker);

                return new UnionPatternMatchCaseGeneratorSmartTagger(buffer, generator) as ITagger<T>;
            }
            
            return null;
        }
    }
}

File: src/FSharpVSPowerTools.Pure/Commands/UnionPatternMatchCaseGeneratorSmartTaggerProvider.fs
module FSharpVSPowerTools.UnionPatternMatchCaseGeneratorSmartTaggerProvider

open System
open System.ComponentModel.Composition
open Microsoft.VisualStudio.Text
open Microsoft.VisualStudio.Text.Editor
open Microsoft.VisualStudio.Text.Operations
open Microsoft.VisualStudio.Text.Tagging
open Microsoft.VisualStudio.Utilities
open Microsoft.VisualStudio.Shell
open FSharpVSPowerTools.Refactoring
open FSharpVSPowerTools.ProjectSystem
open System.Diagnostics
open Microsoft.VisualStudio.Shell.Interop

[<Export (typeof<IViewTaggerProvider>)>]
[<ContentType "F#">]
[<TagType (typeof<UnionPatternMatchCaseGeneratorSmartTag>)>]
type UnionPatternMatchCaseGeneratorSmartTaggerProvider [<ImportingConstructor>]
    ( [<Import(typeof<SVsServiceProvider>)>] 
        serviceProvider             :   IServiceProvider,
        textDocumentFactoryService  :   ITextDocumentFactoryService ,
        undoHistoryRegistry         :   ITextUndoHistoryRegistry    ,
        projectFactory              :   ProjectFactory              ,
        vsLanguageService           :   VSLanguageService           ) =

    interface IViewTaggerProvider with
        member __.CreateTagger (textView, buffer) =
            maybe {
                if textView.TextBuffer <> buffer then return! None else
       //         let! generalOptions = Setting.tryGetGeneralOptions serviceProvider
                let! codeGenOptions = Setting.tryGetCodeGenerationOptions serviceProvider
                let dte = serviceProvider.GetService<EnvDTE.DTE,SDTE>()
                if dte.Version = string VisualStudioVersion.VS2015 then return! None else
                let! doc = textDocumentFactoryService.TryDocumentFromBuffer buffer
                let generator = 
                    new UnionPatternMatchCaseGenerator (doc, textView, undoHistoryRegistry.RegisterHistory buffer, 
                        vsLanguageService, serviceProvider, projectFactory, Setting.getDefaultMemberBody codeGenOptions, OpenDocumentsTracker textDocumentFactoryService )
                return
                    new UnionPatternMatchCaseGeneratorSmartTagger (buffer, generator) :> obj :?> _
            } |> Option.getOrElse null
